pkg/auth/providers/keycloakoidc/keycloak_provider.go

```diff
@@ -175,3 +154,32 @@ func (k *keyCloakOIDCProvider) GetPrincipal(principalID string, token v3.Token)
 	princ := k.toPrincipal(principalType, acct, &token)
 	return princ, err
 }
+
+func (k *keyCloakOIDCProvider) getRefreshAndUpdateToken(ctx context.Context, oauthConfig oauth2.Config, token v3.Token) (*oauth2.Token, error) {
+	var oauthToken *oauth2.Token
+	storedOauthToken, err := k.TokenMGR.GetSecret(token.UserID, token.AuthProvider, []*v3.Token{&token})
+	if err := json.Unmarshal([]byte(storedOauthToken), &oauthToken); err != nil {
+		return oauthToken, err
+	}
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return oauthToken, err
+		}
+		oauthToken.AccessToken = token.ProviderInfo["access_token"]
+	}
+	// Valid will return false if access token is expired
+	if !oauthToken.Valid() {
+		// since token is not valid, the TokenSource func used in the Client func will attempt to refresh the access token
+		// if the refresh token has not expired
+		logrus.Debugf("[generic oidc] RefeshAndUpdateToken: attempting to refresh access token")
+	}
+	reusedToken, err := oauth2.ReuseTokenSource(oauthToken, oauthConfig.TokenSource(ctx, oauthToken)).Token()
+	if err != nil {
+		return oauthToken, err
+	}
+
+	if reflect.DeepEqual(oauthToken, reusedToken) {
```

Comment:
Should this be `!reflect.DeepEqual(oauthToken, reusedToken)` I thought we want to update the token if the new token returned is not same as the one we extracted from the secret?

Reply:
You are correct - I have updated this in my most recent changes.